src/plugins/embeddable/public/lib/panel/panel_header/panel_actions/customize_panel/customize_panel_editor.tsx

```diff
@@ -0,0 +1,305 @@
+ * Side Public License, v 1.
+ */
+
+import React, { useState } from 'react';
+
+import {
+  EuiFormRow,
+  EuiFieldText,
+  EuiSwitch,
+  EuiFlyoutHeader,
+  EuiTitle,
+  EuiFlyoutBody,
+  EuiForm,
+  EuiTextArea,
+  EuiFlyoutFooter,
+  EuiButtonEmpty,
+  EuiButton,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiSuperDatePicker,
+} from '@elastic/eui';
+import { FormattedMessage } from '@kbn/i18n-react';
+import { i18n } from '@kbn/i18n';
+import { TimeRange } from '@kbn/es-query';
+import { hasTimeRange, TimeRangeInput } from './customize_panel_action';
+import { doesInheritTimeRange } from './does_inherit_time_range';
+import { IEmbeddable, Embeddable, EmbeddableOutput, CommonlyUsedRange } from '../../../..';
+
+type PanelSettings = {
+  title?: string;
+  hidePanelTitles?: boolean;
+  description?: string;
+  timeRange?: TimeRange;
+};
+
+interface CustomizePanelProps {
+  embeddable: IEmbeddable;
+  dateFormat?: string;
+  commonlyUsedRanges?: CommonlyUsedRange[];
+  onClose: () => void;
+}
+
+const VISUALIZE_EMBEDDABLE_TYPE = 'visualization';
+
+type VisualizeEmbeddable = IEmbeddable<{ id: string }, EmbeddableOutput & { visTypeName: string }>;
+
+function isVisualizeEmbeddable(
+  embeddable: IEmbeddable | VisualizeEmbeddable
+): embeddable is VisualizeEmbeddable {
+  return embeddable.type === VISUALIZE_EMBEDDABLE_TYPE;
+}
+
+function isTimeRangeCompatible(embeddable: IEmbeddable) {
+  const isInputControl =
+    isVisualizeEmbeddable(embeddable) &&
+    (embeddable as VisualizeEmbeddable).getOutput().visTypeName === 'input_control_vis';
+
+  const isMarkdown =
+    isVisualizeEmbeddable(embeddable) &&
+    (embeddable as VisualizeEmbeddable).getOutput().visTypeName === 'markdown';
+
+  const isImage = embeddable.type === 'image';
+
+  return Boolean(
+    embeddable &&
+      embeddable.parent &&
+      hasTimeRange(embeddable) &&
+      !isInputControl &&
+      !isMarkdown &&
+      !isImage
+  );
+}
+
+export const CustomizePanelEditor = (props: CustomizePanelProps) => {
+  const { onClose, embeddable } = props;
+  const timeRangeCompatible = isTimeRangeCompatible(embeddable);
+  const [hideTitle, setHideTitle] = useState(embeddable.getInput().hidePanelTitles);
+  const [panelDescription, setPanelDescription] = useState(
+    embeddable.getInput().description ?? embeddable.getOutput().defaultDescription
+  );
+  const [panelTitle, setPanelTitle] = useState(
+    embeddable.getInput().title ?? embeddable.getOutput().defaultTitle
+  );
+  const [inheritTimeRange, setInheritTimeRange] = useState(
+    timeRangeCompatible ? doesInheritTimeRange(embeddable as Embeddable<TimeRangeInput>) : false
+  );
+  const [panelTimeRange, setPanelTimeRange] = useState(
+    timeRangeCompatible
+      ? (embeddable as Embeddable<TimeRangeInput>).getInput().timeRange
+      : undefined
+  );
+
+  const commonlyUsedRangesForDatePicker = props.commonlyUsedRanges
+    ? props.commonlyUsedRanges!.map(
```

Comment:
Nit why the `!` here? In the ternary, you're checking if it exists, so hopefully the typescript compiler should know that it exists at this point.